src/merkle/simple_smt/mod.rs

```diff
@@ -0,0 +1,249 @@
+        } else if Self::MAX_DEPTH < depth {
+            return Err(MerkleError::DepthTooBig(depth));
+        }
+
+        let (store, root) = Store::new(depth);
+        let mut tree = Self { root, depth, store };
+        for (key, val) in keys.into_iter().zip(values) {
+            tree.insert_leaf(key, val)
+                .expect("Failed to insert leaf value");
+        }
+
+        Ok(tree)
+    }
+
+    /// Returns the root of this Merkle tree.
+    pub const fn root(&self) -> Word {
+        self.root
+    }
+
+    /// Returns the depth of this Merkle tree.
+    pub const fn depth(&self) -> u32 {
+        self.depth
+    }
+
+    /// Returns a node at the specified key
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified depth is greater than the depth of the tree.
+    /// * The specified key does not exist
+    pub fn get_node(&self, depth: u32, key: u64) -> Result<Word, MerkleError> {
+        if depth == 0 {
+            Err(MerkleError::DepthTooSmall(depth))
+        } else if depth > self.depth() {
+            Err(MerkleError::DepthTooBig(depth))
+        } else if depth == self.depth() {
+            self.store.get_leaf_node(key)
+        } else {
+            let branch_node = self.store.get_branch_node(key, depth)?;
+            Ok(Rpo256::merge(&[branch_node.left, branch_node.right]).into())
+        }
+    }
+
+    /// Returns a Merkle path to the node at the specified key. The node itself is
+    /// not included in the path.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified key does not exist as a branch or leaf node
+    /// * The specified depth is greater than the depth of the tree.
+    pub fn get_path(&self, depth: u32, key: u64) -> Result<Vec<Word>, MerkleError> {
+        if depth == 0 {
+            return Err(MerkleError::DepthTooSmall(depth));
+        } else if depth > self.depth() {
+            return Err(MerkleError::DepthTooBig(depth));
+        }
+
+        let mut path = Vec::with_capacity(depth as usize);
+        let mut curr_key = key;
+        for n in (0..depth).rev() {
+            let parent_key = curr_key >> 1;
+            let parent_node = self.store.get_branch_node(parent_key, n)?;
+            let sibling_node = if curr_key & 1 == 1 {
+                parent_node.left
+            } else {
+                parent_node.right
+            };
+            path.push(sibling_node.into());
+            curr_key >>= 1;
+        }
+        Ok(path)
+    }
+
+    /// Returns a Merkle path to the node at the specified key. The node itself is
+    /// not included in the path.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified key does not exist as a leaf node
+    pub fn get_leaf_path(&self, key: u64) -> Result<Vec<Word>, MerkleError> {
+        if !self.store.check_leaf_node_exists(key) {
+            return Err(MerkleError::InvalidIndex(self.depth, key));
+        }
```

Comment:
The signature of the `get_path(&self, depth: u32, key: u64) -> Result<Vec<Word>, MerkleError>` defines that the user can also fetch nodes. The previous implementation had a bug that it would check for a leaf using the same index, regardless of the depth. Example:

For a tree with depth `3`, if the key `0` is inserted, a `get_path(2, 1)` would fail because the key `1` doesn't exist for the depth `3`, but it exists as a node `(2, 1)`.

The current implementation separated the `get_path` from `get_leaf_path` so the validation that is performed only for the leaves stays with the latter.

We can perform a validation with `get_path`, but it should be something like `if the path from a node to the root is requested for a depth that doesn't belong to the path of a previously inserted leaf, then fail`. But it might not be logical since these "nonexistent" paths will in fact just be the "empty digest" for that level.

Reply:
This makes sense. What I'm thinking is that instead of performing this check in `get_leaf_path()` we can perform it in `get_path()` but only when the `depth` is the same as `self.depth`. Logically, these should be equivalent with the main difference is that we'd also check for empty leaves in `get_path()` when the specified depth indicates that we are getting a leaf node.

Reply:
Sounds good!